parsl/executors/high_throughput/executor.py

```diff
@@ -0,0 +1,397 @@
+                                  args=(comm_q,),
+                                  kwargs={"client_ports": (self.outgoing_q.port, self.incoming_q.port),
+                                          "worker_ports": self.worker_ports,
+                                          "worker_port_range": self.worker_port_range
+                                  },
+        )
+        self.queue_proc.start()
+        try:
+            (worker_task_port, worker_result_port) = comm_q.get(block=True, timeout=120)
+        except queue.Empty:
+            logger.error("Interchange has not completed initialization in 120s. Aborting")
+            raise Exception("Interchange failed to start")
+
+        self.worker_task_url = "tcp://{}:{}".format(self.public_ip, worker_task_port)
+        self.worker_result_url = "tcp://{}:{}".format(self.public_ip, worker_result_port)
+
+    def _start_queue_management_thread(self):
+        """Method to start the management thread as a daemon.
+
+        Checks if a thread already exists, then starts it.
+        Could be used later as a restart if the management thread dies.
+        """
+        if self._queue_management_thread is None:
+            logger.debug("Starting queue management thread")
+            self._queue_management_thread = threading.Thread(target=self._queue_management_worker)
+            self._queue_management_thread.daemon = True
+            self._queue_management_thread.start()
+            logger.debug("Started queue management thread")
+
+        else:
+            logger.debug("Management thread already exists, returning")
+
+    def submit(self, func, *args, **kwargs):
+        """Submits work to the the outgoing_q.
+
+        The outgoing_q is an external process listens on this
+        queue for new work. This method behaves like a
+        submit call as described here `Python docs: <https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor>`_
+
+        Args:
+            - func (callable) : Callable function
+            - *args (list) : List of arbitrary positional arguments.
+
+        Kwargs:
+            - **kwargs (dict) : A dictionary of arbitrary keyword args for func.
+
+        Returns:
+              Future
+        """
+        task_id = uuid.uuid4()
+
+        logger.debug("Pushing function {} to queue with args {}".format(func, args))
+
+        self.tasks[task_id] = Future()
+
+        fn_buf = pack_apply_message(func, args, kwargs,
+                                    buffer_threshold=1024 * 1024,
+                                    item_threshold=1024)
+
+        msg = {"task_id": task_id,
+               "buffer": fn_buf}
+
+        # Post task to the the outgoing queue
+        self.outgoing_q.put(msg)
+
+        # Return the future
+        return self.tasks[task_id]
+
+    @property
+    def scaling_enabled(self):
+        return self._scaling_enabled
+
+    def scale_out(self, workers=1):
```

Comment:
Why is `scale_out` in units of workers and `scale_in` in units of blocks? Could we make them the same?

Reply:
Replaces workers with blocks